### ISSUE TITLE
Kernel not switched after two switches

### ISSUE THREAD

**Issue by buis12** [microsoft/vscode-jupyter#4312]
Issue Type: <b>Bug</b>

1. Within a Jupyter notebook switch kernel (works)
2. Again switch the kernel (works)
3. Again switch the kernel (fails)

In the logs the following event doesn't appear for step 3:
Started kernel Python 3.9.1 64-bit ('env_name1': conda)

Using conda environments with python 3.8.3, 3.8.6 and 3.9.1.

Extension version: 2020.12.414227025
VS Code version: Code 1.52.1 (ea3859d4ba2f3e577a159bc91e3074c5d85c0523, 2020-12-16T16:34:46.910Z)
OS version: Windows_NT x64 10.0.19042

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz (8 x 1992)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>opengl: enabled_on<br>protected_video_decode: enabled<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.87GB (5.33GB free)|
|Process Argv|--crash-reporter-id 450a0805-b0f4-478f-a3f5-77ec6b65155e|
|Screen Reader|no|
|VM|0%|
</details>

**Comment by DonJayamanne** [comment]
Thanks for filing this issue,please could you  provide the logs:

Please could you enable logging as follows:
* Open settings & go into `Jupyter->Logging`
* Change the value to `debug`
* Reload VS Code,
* Run step 1, and capture the logs from the output panel
* Next clear the output panel
* Run step 2, and capture the logs from the output panel
* Next clear the output panel
* Finally Run step 3, and capture the logs from the output panel
* Next clear the output panel

Please upload the 3 log files. I'd like to have the three separately so we can determine when you perform each step, also its possible there are errors logged which could indicate the cause of the issue.

**Questions:**
* What kernels are you switching from & to in each step (please provide the names, so I can look for it in the logs).
* What version of conda are you using?

**Comment by buis12** [comment]
Thanks for the response. 

I've added the **logs** as attachement:
- <username> is replaced by User_Name
- <username>/<onedrive_name> is replaced by PATH

Conda version: **conda 4.9.2**

I'm switching between **kernels** 1 and 2.
Kernel 1: 
'C:\\Users\\User_Name\\Anaconda3\\python.exe'
Python 3.8.3 64-bit (conda)

Kernel 2: 
'C:\\Users\\User_Name\\Anaconda3\\envs\\scrapy1\\python.exe'
Python 3.8.6 64-bit ('scrapy1': conda)

Steps of switching:
1. VScode starts using Kernel 1.
2. I try to switch to Kernel 2.
3. I try to switch back to Kernel 1.

Kernels of the different steps (based on output of the notebook):
1. 'C:\\Users\\User_Name\\Anaconda3\\python.exe'
2. 'C:\\Users\\User_Name\\Anaconda3\\envs\\scrapy1\\python.exe'
3. 'C:\\Users\\User_Name\\Anaconda3\\envs\\scrapy1\\python.exe'[](url)


[vscode_switchkernels1.txt](https://github.com/microsoft/vscode-jupyter/files/5781277/vscode_switchkernels1.txt)
[vscode_switchkernels2.txt](https://github.com/microsoft/vscode-jupyter/files/5781275/vscode_switchkernels2.txt)
[vscode_switchkernels3.txt](https://github.com/microsoft/vscode-jupyter/files/5781276/vscode_switchkernels3.txt)

**Comment by DonJayamanne** [comment]
Hmm, please could you try the following:
* When you have gone to step 1, save the notebook, close the notebook & reload VS Code
* Re-open that same notebook & run a cell & confirm `scrapy1` is the the selected kernel (it should be)
* Next try to switch to the base kernel

**Comment by buis12** [comment]
This works. Even without saving the notebook. The set kernel, even if it didn't switch correctly, but was selected before closing will load when restarting VS Code. (assuming reloading and restarting means the same)

Switching in both direction works. It's really about the third kernel, or in other words the second switch.

**Comment by DonJayamanne** [comment]
> Switching in both direction works. It's really about the third kernel, or in other words the second switch.

So let me get this right:
* Open notebook select scrapy, run cell & confirm its ok
* Close notebook
* Reload/restart VS Code (either one)
* Open same notebook & run cell, confirm scrap is the current kernel
* Switch to base conda environment, confirm this works
* Switch back to scrap (2nd switch) & this does not work, please confirm
Thanks

**Comment by buis12** [comment]
That is correct.

**Comment by DonJayamanne** [comment]
@buis12 apologies for silence from my end.
We're making a few changes to this area of the code & hope it will get resolved soon.
In the mean time please could you try the following:
* Install VS Code insiders
* Install the Jupyter extension
* Try it in the new notebook & let me know if you can repro it there as well.
* Oh yes, please enable logging & provide the logs (for each step, as previously done, thanks).
FYI - The Notebook UI is different in VS Code insiders (this is a preview notebook that we're moving to).

**Comment by buis12** [comment]
Somehow this doesn't work as well.

Steps I've taken:

1. Installing VS Code insiders
2. Installing python extension
3. Installing jupyter extension
4. Open the testSwitchingKernels.ipynb.
5. Run the first cell to check the sys.executable, copy and clear debug
6. Switch kernel using ctrl+shift+p: Jupyter select interpreter to start jupyter server
7. Run the first cell again, copy and clear debug
8. Switch kernel using ctrl+shift+p: Jupyter select interpreter to start jupyter server
9. Run the first cell again, copy and clear debug

In this case, the first kernel switch already fails. So the second "switch" is not a real switch. Every run I got the location of Kernel 1.

I'm starting in kernel 1, trying to switch to kernel 2, trying to switch back to kernel 1.
Kernel 1:
'C:\Users\User_Name\Anaconda3\python.exe'
Python 3.8.3 64-bit (conda)

Kernel 2:
'C:\Users\User_Name\Anaconda3\envs\scrapy1\python.exe'
Python 3.8.6 64-bit ('scrapy1': conda)

[vscodeinsiders_switchkernel1.txt](https://github.com/microsoft/vscode-jupyter/files/5874103/vscodeinsiders_switchkernel1.txt)
[vscodeinsiders_switchkernel2.txt](https://github.com/microsoft/vscode-jupyter/files/5874102/vscodeinsiders_switchkernel2.txt)
[vscodeinsiders_switchkernel3.txt](https://github.com/microsoft/vscode-jupyter/files/5874104/vscodeinsiders_switchkernel3.txt)

**Comment by DonJayamanne** [comment]
> Switch kernel using ctrl+shift+p: Jupyter select interpreter to start jupyter server

Thanks for the feedback.
Looks like you are using the incorrect command to switch the kernel.
The command `Jupyter select interpreter to start jupyter server`, doesn't change the kernel. This is used to start Jupyter if we're unable to start a notebook without running Jupyter itself (a fallback approach).

To switch kernels, please use the command `Jupyter: Select a Kernel`
If you are in a Notebook, please use the UI

Closing the issue for now as the command used was incorrect.
If the issue persists, please feel free to comment on this and I'll be able to re-open the issue.

**Comment by buis12** [comment]
Thanks for pointing out my mistake. Now I've tried the command Jupyter: Select a Kernel and the UI option (bottom right). This solves the problem for VSCode insider. Before I couldn't find the UI option because I was used to it being in the top right corner.

Just to be clear, the problem in (the old) VSCode (still) exists. I've performed the kernel switch using the UI and now I've also tried `Jupyter: Select a Kernel`. Both methods fail switching the second time. Also, I haven't try an reinstall yet because we were hunting bugs.
